### PR TITLE
CB-9900 Determine if Subnet is Private or Public for Load Balancer.

### PR DESCRIPTION
**Pull request by orngejaket**
This deterines based on the subnet's route tables if there is an internet
gateway attached to it making it either a private or public load balancer.

See detailed description in the commit message.

### CODE REVIEW

**Review by hreeve-cloudera — APPROVED** [review]

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsLaunchService.java [hunk 1]
```diff
@@ -249,10 +254,16 @@ private void updateCloudformationWithLoadBalancers(AuthenticatedContext ac, Clou
         }
     }
 
-    private AwsLoadBalancer convert(CloudLoadBalancer cloudLoadBalancer, CloudStack stack, List<CloudResource> instances) {
+    private AwsLoadBalancer convert(CloudLoadBalancer cloudLoadBalancer, CloudStack stack,
+            List<CloudResource> instances, Network network, AmazonEC2Client amazonEC2Client) {
+
+        String subnetId = network.getStringParameter("subnetId");
```
[the remark this answers is not in the document]
Comment (lacikaaa) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsLaunchService.java:260]: actually, we have `AwsNetworkView` in the caller method, so if you pass it as an argument, or just the subnetId itself as the network is not used for anything else you could omit this part

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsLaunchService.java [hunk 2]
```diff
@@ -249,10 +254,16 @@ private void updateCloudformationWithLoadBalancers(AuthenticatedContext ac, Clou
...
+        DescribeRouteTablesResult describeRouteTablesResult = amazonEC2Client.describeRouteTables();
+        AwsLoadBalancerScheme scheme = awsSubnetIgwExplorer.hasInternetGatewayOfSubnet(describeRouteTablesResult, subnetId)
+                ? AwsLoadBalancerScheme.PUBLIC : AwsLoadBalancerScheme.PRIVATE;
```
Comment (lacikaaa) [inline cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsLaunchService.java:263]: this scheme creation could be in a separate method

**Review by lacikaaa — CHANGES_REQUESTED** [review]
Honestly, I was going to ask about the unit test for it, but the whole new LB stuff is missing unit tests :(

**Comment by hreeve-cloudera** [comment]
> Honestly, I was going to ask about the unit test for it, but the whole new LB stuff is missing unit tests :(

@lacikaaa The LB changes were introduced in https://github.com/hortonworks/cloudbreak/pull/9475, and included a number of new unit tests. However, neither Steve nor I could tell exactly how `AwsLaunchService` was being tested, so we weren't sure where to add tests for this specific class.

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsLaunchService.java [hunk 3]
```diff
@@ -249,10 +254,19 @@ private void updateCloudformationWithLoadBalancers(AuthenticatedContext ac, Clou
         }
     }
 
-    private AwsLoadBalancer convert(CloudLoadBalancer cloudLoadBalancer, CloudStack stack, List<CloudResource> instances) {
+    private AwsLoadBalancerScheme determinePublicVsPrivateSchema(AwsNetworkView awsNetworkView, AmazonEC2Client amazonEC2Client) {
+        String subnetId = awsNetworkView.getExistingSubnet();
```
Comment (lacikaaa) [inline cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsLaunchService.java:258]: are we sure it's not a subnet list here? shall we validate?

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsLaunchService.java [hunk 4]
```diff
@@ -249,10 +254,19 @@ private void updateCloudformationWithLoadBalancers(AuthenticatedContext ac, Clou
...
+        DescribeRouteTablesResult describeRouteTablesResult = amazonEC2Client.describeRouteTables();
+        return awsSubnetIgwExplorer.hasInternetGatewayOfSubnet(describeRouteTablesResult, subnetId)
+                ? AwsLoadBalancerScheme.PUBLIC : AwsLoadBalancerScheme.PRIVATE;
+    }
+
+    private AwsLoadBalancer convert(CloudLoadBalancer cloudLoadBalancer, List<CloudResource> instances,
```
Comment (lacikaaa) [inline cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsLaunchService.java:264]: could you give it a better name?

**Comment by lacikaaa** [comment]
> However, neither Steve nor I could tell exactly how AwsLaunchService was being tested, so we weren't sure where to add tests for this specific class.

@hreeve-cloudera Please file a followup jira for the missing tests and add them. If you need any help figuring out something we are happy to help.

**Comment by hreeve-cloudera** [comment]
> > However, neither Steve nor I could tell exactly how AwsLaunchService was being tested, so we weren't sure where to add tests for this specific class.
> 
> @hreeve-cloudera Please file a followup jira for the missing tests and add them. If you need any help figuring out something we are happy to help.

Opened https://jira.cloudera.com/browse/CB-10280. I think I've identified AwsLaunchTest as where we need to add new tests. @lacikaaa let me know if you think we're missing anything else.

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsLaunchService.java (hunk 3 above)
Comment (orngejaket) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsLaunchService.java:258]: In debugging, I have only seen 1 subnet get passed in when I select 3 during env creation.

Comment (orngejaket) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsLaunchService.java:258]: Hreeve is working on a change for semi-private networks to support a list of subnets.

**Review by lacikaaa — APPROVED** [review]

**Pull request merged**